main/ertp/api/amount-math.md

```diff
@@ -1,189 +1,304 @@
+
+- `MathKind.NAT` (`nat`): Used with fungible assets. `amount` `values` are natural numbers (non-negative integers).
+- `MathKind.STRING_SET` (`strSet`): Used with non-fungible assets. `amount` `values` are strings.
+- `MathKind.SET` (`set`): Used with non-fungible assets. `amount` `values` are objects or records with multiple properties.
+
+Use `makeIssuerKit(allegedName, MathKind)` to specify which `amountMath` 
+kind your contract uses. The second parameter, `MathKind` is optional and 
+defaults to `MathKind.NAT` if not given. For example
+```js
+import { MathKind, makeIssuerKit } from '@agoric/ertp';
+makeIssuerKit('quatloos`); // Defaults to 'MathKind.NAT'
+makeIssuerKit('foobars', 'MathKind.STRING_SET');
+makeIssuerKit('kitties', MathKind.SET');
+```
 
 ## Amount
 
-Amounts are descriptions of digital assets, answering the questions "how much" and "of what kind". Amounts are values labeled with a brand. AmountMath executes the logic of how amounts are changed when digital assets are merged, separated, or otherwise manipulated. For example, a deposit of 2 bucks into a purse that already has 3 bucks gives a new balance of 5 bucks. An empty purse has 0 bucks. AmountMath relies heavily on polymorphic MathHelpers, which manipulate the unbranded portion.
+An `amount` is a description of digital assets, answering the questions "how much?" and "of what kind?". It is a `value` ("how much") labeled with a `brand` ("of what kind"). `AmountMath` executes the logic of how an `amount` changes when digital assets are merged, separated, or otherwise manipulated. For example, a deposit of 2 Quatloos into a `purse` that already has 3 Quatloos gives a new balance of 5 Quatloos. An empty `purse` has 0 Quatloos. 
 
 ```js
 someAmount: {
-  brand,
-  value: someValue
+  brand: someBrand,
+  value: someValue,
 }
 ```
 
 ## Value
 
-Values describe how much of something that can be owned or shared. Fungible values are normally represented by natural numbers. Other values may be represented as strings naming a particular right, or an arbitrary object that sensibly represents the rights at issue.
+`values` describe how much of something can be owned or shared. A fungible `value` is normally represented by a natural number. Other `values` may be represented as strings naming a particular right, or an arbitrary object that sensibly represents the rights at issue.
+
+A `value` must be `Comparable`.
+
+## LocalAmountMath
+
+We encourage you to make and use `LocalAmountMath`, a local and thus synchronous version of `AmountMath`. 
+Their local or remote status is the only different between the two types; each has the same methods, the
+same kinds (`MathKind.NAT`, etc.) and the same one-to-one relationship with a `mint`, `issuer`, and `brand`.
+If necessary, both a `LocalAmountMath` and an `AmountMath` can exist and be associated with the same
+`mint`, `issuer`, and `brand` **tyg todo This last is a guess on my part. Correct?**.
```

Comment:
This is mostly a description for TYG, and not intended as text for this doc.

local AmountMaths aren't a different type from remote ones, they merely reside in the same vat as the user. This allows them to be accessed without remote calls. Both use the same code, and hold the same state. The Brand is a unique object that has a unique instance somewhere in the distributed object system. 

The way we share objects and references across the networks distinguishes things with identity and copyable values. by-copy values include numbers and strings. Every instance of 3 is indistinguishable, so we make copies when we need to, and one can't tell whether any particular '3' is the same as another. OTOH, objects are unique and live in only one place. If an object in one vat has a reference to another object from another vat, we represent that with a local object called a "presence" that can forward messages to the original. *We ensure that there is only one presence for any particular object in each vat* that knows about it. This means if there are two different objects in vat A that both have references to some object in vat B, they will be the same object.

This unique object is what we use for the Brand held by an AmountMath. The rest of the state of the AmountMath is copyable state. When you have a local AmountMath, it's holding onto the presence for the remote Brand, and so can recognize any Amount that is from the same issuer. A remote AmountMath has the same ability to distinguish between Amounts that came from that Issuer and others because as the Amounts are shared around, their values are copied, but the Brands they contain are represented as presences, and so can be uniquely identified in every vat in which they appear. 

The only reason one doesn't want to use a remote AmountMath is that every call takes a network roundtrip. We designed the representation of AmountMath so a local copy is every bit as good as a remote one, and faster to access.